libraries/standard/mqtt/src/mqtt.c

```diff
@@ -1377,20 +1377,29 @@ MQTTStatus_t MQTT_Init( MQTTContext_t * pContext,
         LogError( ( "Argument cannot be NULL: pContext=%p, "
                     "pTransportInterface=%p, "
                     "pNetworkBuffer=%p",
-                    pContext,
-                    pTransportInterface,
-                    pNetworkBuffer ) );
+                    ( void * ) pContext,
+                    ( void * ) pTransportInterface,
+                    ( void * ) pNetworkBuffer ) );
         status = MQTTBadParameter;
     }
-    else if( ( getTimeFunction == NULL ) || ( userCallback == NULL ) ||
-             ( pTransportInterface->recv == NULL ) || ( pTransportInterface->send == NULL ) )
+    else if( getTimeFunction == NULL )
     {
-        LogError( ( "Function pointers cannot be NULL: getTimeFunction=%p, userCallback=%p, "
-                    "transportRecv=%p, transportRecvSend=%p",
-                    getTimeFunction,
-                    userCallback,
-                    pTransportInterface->recv,
-                    pTransportInterface->send ) );
+        LogError( ( "Invalid parameter: getTimeFunction is NULL" ) );
```

Comment:
What's the purpose in separating these parameters? It seems more repetitive this way, and it doesn't match the condition earlier in this function, at line 1374

Reply:
The earlier implementation was printing the function pointers under a single if condition so that the reader could detect which of the function pointers is null. However, casting function pointers to `void *` for printing is not supported by ISO C.
GCC gives the following warning when casting function pointers to `void *`
```
/home/ubuntu/Repos/aws-iot-device-sdk-embedded-C/libraries/standard/mqtt/src/mqtt.c:1387:21: warning: format ‘%p’ expects argument of type ‘void *’, but argument 2 has type ‘MQTTGetCurrentTimeFunc_t {aka unsigned int (*)(void)}’ [-Wformat=]
         LogError( ( "Invalid parameter: getTimeFunction is NULL: %p", getTimeFunction ) );
                     ^
/home/ubuntu/Repos/aws-iot-device-sdk-embedded-C/demos/logging-stack/logging_stack.h:49:40: note: in definition of macro ‘SdkLog’
     #define SdkLog( string )    printf string
                                        ^~~~~~
/home/ubuntu/Repos/aws-iot-device-sdk-embedded-C/libraries/standard/mqtt/src/mqtt.c:1387:9: note: in expansion of macro ‘LogError’
         LogError( ( "Invalid parameter: getTimeFunction is NULL: %p", getTimeFunction ) );
```